### PR TITLE
Changing construction of the first guess and coefficient matrices in NLTE solver

### PR DESCRIPTION
**Pull request by sonachitchyan**
### :pencil: Description

**Type:** :beetle: `bugfix` | :rocket: `feature` | :biohazard: `breaking change` | :vertical_traffic_light: `testing` | :memo: `documentation` | :roller_coaster: `infrastructure`

This PR makes sure that thewayt coefficient matrices and the first guess of NLTE solver are constructed is able to work both with and without NLTE excitation species.

Also, link issues affected by this pull request by using the keywords: `close`, `closes`, `closed`, `fix`, `fixes`, `fixed`, `resolve`, `resolves` or `resolved`. 

### :pushpin: Resources

Examples, notebooks, and links to useful references.

### :vertical_traffic_light: Testing

How did you test these changes?

- [ ] Testing pipeline
- [ ] Other method (describe)
- [ ] My changes can't be tested (explain why)

### :ballot_box_with_check: Checklist

- [ ] I requested two reviewers for this pull request
- [ ] I updated the documentation according to my changes
- [ ] I built the documentation by applying the `build_docs` label

> **Note:** If you are not allowed to perform any of these actions, ping (@) a contributor.

### CODE REVIEW

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 1]
```diff
@@ -699,3 +767,170 @@ def prepare_solution_vector(self, number_density):
...
+        r_lu_matrix_reshaped[r_lu_index] = (
+            B_lus[np.newaxis].T * j_blues_filtered * beta_sobolev_filtered
+        )
+
+        rates_matrix_bound_bound = r_lu_matrix + r_ul_matrix
+        for i in range(number_of_levels):
+            rates_matrix_bound_bound[i, i] = -rates_matrix_bound_bound[
+                :, i
+            ].sum(axis=0)
+        rates_matrix_bound_bound[0, :, :] = 1.0
```
Comment (sonachitchyan) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:889]: @chvogl do you know why we need this line?

File: tardis/plasma/properties/nlte_excitation_data.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import logging
+
+import numpy as np
+
+logger = logging.getLogger(__name__)
+
+
+class NLTEExcitationData(object):
+    def __init__(self, atom_data_lines, nlte_excitation_species):
```
Comment (Rodot-) [inline tardis/plasma/properties/nlte_excitation_data.py:9]: Needs docstrings

File: tardis/plasma/properties/nlte_excitation_data.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        self.atom_data_lines = atom_data_lines
+        self.lines = atom_data_lines.reset_index()
+        self.nlte_excitation_species = nlte_excitation_species
+
+        if nlte_excitation_species:
```
Comment (Rodot-) [inline tardis/plasma/properties/nlte_excitation_data.py:14]: perhaps nlte_excitation_species should be a kwarg set to None/False?

**Review by Rodot- — CHANGES_REQUESTED** [review]
It looks like you are currently missing some units tests for your individual functions.  Might be good to include quick simple tests that make sure the code works, for example, with both empty and non-empty nlte_excitation_species lists.

File: tardis/plasma/properties/nlte_excitation_data.py (hunk 3 above)
Comment (sonachitchyan) [reply tardis/plasma/properties/nlte_excitation_data.py:14]: It's default is empty either way, and I will be calling this function only when nlte_excitation_species is not empty

File: tardis/plasma/properties/nlte_excitation_data.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
+import logging
+
+import numpy as np
+
+logger = logging.getLogger(__name__)
+
+
+class NLTEExcitationData(object):
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_excitation_data.py:8]: Most of this class is code from the `NLTEData` class in `io/atom_data/base.py`. Is it possible to minimize the code duplication by sharing some of the methods?

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 5]
```diff
@@ -1,8 +1,10 @@
 import pandas as pd
 import numpy as np
 from scipy.optimize import root
+from tardis.io.atom_data.base import NLTEData
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:4]: This imported but not used?

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 6]
```diff
@@ -25,6 +27,7 @@ def calculate(
         phi,
         rate_matrix_index,
         number_density,
+        nlte_excitation_species,
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:30]: This is not added to the parameters in the docstring.

**Review by chvogl — COMMENTED** [review]
I only managed to look at the first part. I will continue when I have time.

**Comment by sonachitchyan** [comment]
> It looks like you are currently missing some units tests for your individual functions. Might be good to include quick simple tests that make sure the code works, for example, with both empty and non-empty nlte_excitation_species lists.

planning on doing this after the whole solver is in. So that I can also test the construction of the matrices as a whole.

**Review by atharva-2001 — DISMISSED** [review]

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 7]
```diff
@@ -573,13 +643,12 @@ def prepare_first_guess(
...
         """
-        # TODO needs to be changed for excitation
-        array_size = (number_density.index.values + 1).sum() + 1
-        first_guess = np.zeros(array_size)
-        index = 1
+        first_guess = pd.Series(0.0, index=rate_matrix_index)
         for atomic_number in atomic_numbers:
-            first_guess[index] = number_density.loc[atomic_number]
-            index += atomic_number + 1
+            first_guess.loc[(atomic_number, 1)][0] = number_density.loc[
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:648]: In principle, after the first iteration we should use the level number densities from the previous iteration as a first guess. For now, maybe adding a TODO note is sufficient.

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 8]
```diff
@@ -699,3 +768,168 @@ def prepare_solution_vector(self, number_density):
...
+        beta_sobolev : Numpy Array, dtype float
+            Sobolev escape probability
+        excitation_species : tuple
+            Species treated in NLTE excitation.
+        nlte_data : NLTEExcitationData
+            Data relevant to NLTE exciation species.
+
+        Returns
+        -------
+        numpy.array
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:842]: Add shape information.

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 9]
```diff
@@ -699,3 +768,168 @@ def prepare_solution_vector(self, number_density):
...
+        t_electrons :  Numpy Array, dtype float
+            Electron temperatures.
+        j_blues : Pandas DataFrame, dtype float
+            J_blue values as calculated in LTE.
+        beta_sobolev : Numpy Array, dtype float
+            Sobolev escape probability
+        excitation_species : tuple
+            Species treated in NLTE excitation.
+        nlte_data : NLTEExcitationData
+            Data relevant to NLTE exciation species.
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:838]:
```suggestion
            Data relevant to NLTE excitation species.
```

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 10]
```diff
@@ -699,3 +768,168 @@ def prepare_solution_vector(self, number_density):
...
+        )
+        coeff_matrix = pd.DataFrame(
+            coeff_array, columns=coeff_matrix_without_exc.columns, index=index
+        )
+        return coeff_matrix
+
+    @staticmethod
+    def main_nlte_calculation_bound_bound(
+        atomic_data_levels,
+        t_electrons,
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:817]: This doesn't really need `t_electrons`. It only needs number of shells.

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 11]
```diff
@@ -699,3 +768,168 @@ def prepare_solution_vector(self, number_density):
...
+        """Calculates a matrix with bound-bound rates for NLTE excitation treatment.
+
+        Parameters
+        ----------
+        atomic_data : Object
+            Atomic data from the atomic datafile.
+        t_electrons :  Numpy Array, dtype float
+            Electron temperatures.
+        j_blues : Pandas DataFrame, dtype float
+            J_blue values as calculated in LTE.
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:832]:
```suggestion
            Mean intensities in the blue wings of the line transitions.
```

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 12]
```diff
@@ -699,3 +768,168 @@ def prepare_solution_vector(self, number_density):
...
+        beta_sobolev : Numpy Array, dtype float
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:833]:
```suggestion
        beta_sobolev : numpy.array, dtype float
```

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 13]
```diff
@@ -699,3 +768,168 @@ def prepare_solution_vector(self, number_density):
...
+        )
+
+        rates_matrix_bound_bound = r_lu_matrix + r_ul_matrix
+        for i in range(number_of_levels):
+            rates_matrix_bound_bound[i, i] = -rates_matrix_bound_bound[
+                :, i
+            ].sum(axis=0)
+        return rates_matrix_bound_bound
+
+        # TODO: beta sobolev needs to be recalculated for each iteration, because it depends on numberdensity
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:892]: All of the stuff that uses the beta_sobolevs should be moved to a different method that only works on a single shell at a time. We want to solve the non-linear equations for each shell separately, which requires us to recalculate the bound-bound rate matrix with new values of beta_sobolev at every evaluation of the solver. The current method is still useful since we can pre-calculate a lot of the things we need.

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 14]
```diff
@@ -699,3 +768,168 @@ def prepare_solution_vector(self, number_density):
...
+
+    def calculate_coll_exc_coefficients(
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:894]: Is this really needed? Can you not just do an indexing operation later when you loop over the nlte_excitation_species to build the rate matrices?

**Review by chvogl — CHANGES_REQUESTED** [review]
I still need to look at e.g. prepare_coefficient_matrices_excitation and the tests, but here is some initial feedback.

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 15]
```diff
@@ -438,6 +449,61 @@ def prepare_ion_recomb_coefficients_nlte_ion(
             .groupby(level=("atomic_number", "ion_number"))
             .sum()
         )
+        if len(nlte_excitation_species) != 0:
+            total_photo_ion_coefficients_with_levels = (
+                (level_population_fraction.loc[gamma.index] * gamma)
```
[the remark this answers is not in the document]
Comment (sonachitchyan) [reply tardis/plasma/properties/nlte_rate_equation_solver.py:454]: yeah I misunderstood that, should I fix this for the other rates too?

File: tardis/plasma/properties/nlte_rate_equation_solver.py (hunk 13 above)
Comment (sonachitchyan) [reply tardis/plasma/properties/nlte_rate_equation_solver.py:894]: When calling this, I was thinking of calling it for the individual shell(as our current solver also works that way for NLTE ionization treatment). Should I still move the stuff connected to beta_sobolevs to another method?

File: tardis/plasma/properties/nlte_rate_equation_solver.py (hunk 14 above)
Comment (sonachitchyan) [reply tardis/plasma/properties/nlte_rate_equation_solver.py:896]: I was thinking of getting the rates once and just using them afterwards, to be consistent with other rates that we have

File: tardis/plasma/properties/nlte_excitation_data.py (hunk 4 above)
Comment (wkerzendorf) [reply tardis/plasma/properties/nlte_excitation_data.py:8]: issue

File: tardis/plasma/properties/nlte_rate_equation_solver.py (hunk 13 above)
Comment (chvogl) [reply tardis/plasma/properties/nlte_rate_equation_solver.py]: The way it is coded it is intended to be called with multiple shells, see, e.g.:
`r_lu_matrix_reshaped = r_lu_matrix.reshape((number_of_levels**2, len(t_electrons)))`
You can pass in `len(t_electrons) = 1` but that is not the best solution.

It is not crucial to move the stuff connected to beta_sobolevs to another method but it would save computational time since we can skip half of the steps in setting up the rate matrix (which we will need to do a lot).

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 16]
```diff
@@ -699,3 +770,186 @@ def prepare_solution_vector(self, number_density):
...
+            if treated in NLTE ionization, 3rd index is "nlte_ion".
+        coeff_matrix_without_exc : pandas.DataFrame
+            Dataframe of rates summed over ion numbers.
+        coeff_matrix_with_exc : pandas.DataFrame
+            Dataframe of rates without summing over the ion number.
+
+        Returns
+        -------
+        pandas.DataFrame
+            Returns a combines dataframe of coefficients, with NLTE excitation treatment taken into account.
```
Comment (wkerzendorf) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:796]:
```suggestion
            Returns a combined dataframe of coefficients, with NLTE excitation treatment taken into account.
```

File: tardis/plasma/properties/nlte_rate_equation_solver.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (sonachitchyan) [reply tardis/plasma/properties/nlte_rate_equation_solver.py]: done

File: tardis/plasma/properties/nlte_excitation_data.py (hunk 4 above)
Comment (chvogl) [reply tardis/plasma/properties/nlte_excitation_data.py:8]: Described in #2266.

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 17]
```diff
@@ -699,3 +711,202 @@ def prepare_solution_vector(self, number_density):
...
+        Parameters
+        ----------
+        rate_matrix_index : MultiIndex
+            (atomic_number, ion_number, treatment type)
+            If ion is treated in LTE or nebular ionization, 3rd index is "lte_ion",
+            if treated in NLTE ionization, 3rd index is "nlte_ion".
+        coeff_matrix_without_exc : pandas.DataFrame
+            Dataframe of rates summed over ion numbers.
+        coeff_matrix_with_exc : pandas.DataFrame
+            Dataframe of rates without summing over the ion number.
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:732]: Does this include collisional and radiative bound-bound rates?

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 18]
```diff
@@ -699,3 +711,202 @@ def prepare_solution_vector(self, number_density):
...
+
+        Returns
+        -------
+        pandas.DataFrame
+            Returns a combined dataframe of coefficients, with NLTE excitation treatment taken into account.
+        """
+        coeff_array = np.zeros(
+            (rate_matrix_index.size, coeff_matrix_without_exc.shape[1])
+        )
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:741]: Isn't this overwritten in line 747?

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 19]
```diff
@@ -699,3 +711,202 @@ def prepare_solution_vector(self, number_density):
             )
         solution_vector = np.hstack(solution_array + [0])
         return solution_vector
+
+    @staticmethod
+    def prepare_coefficient_matrices_excitation(
```
Comment (chvogl) [inline tardis/plasma/properties/nlte_rate_equation_solver.py:716]: I don't quite understand what this method does. It would be useful to expand upon the description in the docstring (including e.g. shape information for the created DataFrame).

**Review by chvogl — CHANGES_REQUESTED** [review]
I still think it would be good to refactor prepare_bound_bound_rate_matrix.

File: tardis/plasma/properties/nlte_rate_equation_solver.py (hunk 14 above)
Comment (sonachitchyan) [reply tardis/plasma/properties/nlte_rate_equation_solver.py:855]: I can get rid of it and construct it "on the go" while preparing the rate matrix.

File: tardis/plasma/tests/test_nlte_excitation.py [hunk 20]
```diff
@@ -0,0 +1,92 @@
...
+        copy_atomic_dataset.lines, simple_excitation_species
+    )
+    simple_number_of_shells = 1
+    simple_j_blues = pd.DataFrame(
+        0.5, index=copy_atomic_dataset.lines.index, columns=["0"]
+    )
+
+    simple_beta_sobolev = pd.DataFrame(
+        0.8, index=copy_atomic_dataset.lines.index, columns=["0"]
+    )
```
[the remark this answers is not in the document]
Comment (sonachitchyan) [reply tardis/plasma/tests/test_nlte_excitation.py]: done!

File: tardis/plasma/tests/test_nlte_excitation.py [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+    NLTERateEquationSolver,
+)
+from tardis.plasma.properties.nlte_excitation_data import NLTEExcitationData
+
+
+def test_prepare_bound_bound_rate_matrix(
+    nlte_atomic_dataset,
+):
+    """
+    Using a simple case of nlte_ion for HI and HeII, checks if the calculate_rate_matrix generates the correct data.
```
Comment (chvogl) [inline tardis/plasma/tests/test_nlte_excitation.py:16]:
```suggestion
    Using a simple case of nlte_exc for HI, checks if prepare_bound_bound_rate_matrix generates the correct data.
```

**Review by chvogl — COMMENTED** [review]
This already looks really good! We should be able to merge this after some minor changes.

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 22]
```diff
@@ -699,3 +713,137 @@ def prepare_solution_vector(self, number_density):
...
+        try:
+            beta_sobolev_filtered = beta_sobolev.iloc[lines_index]
+        except AttributeError:
+            beta_sobolev_filtered = beta_sobolev
+        r_ul_matrix_reshaped = r_ul_matrix.reshape(
+            (number_of_levels**2, number_of_shells)
+        )
+        r_lu_matrix_reshaped = r_lu_matrix.reshape(
+            (number_of_levels**2, number_of_shells)
+        )
```
[the remark this answers is not in the document]
Comment (sonachitchyan) [reply tardis/plasma/properties/nlte_rate_equation_solver.py:760]: I return the not reshaped version from `prepare_r_uls_r_lus`

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 23]
```diff
@@ -699,3 +713,137 @@ def prepare_solution_vector(self, number_density):
...
+            Matrix with the rates(lower to upper transition) of bound-bound interaction(DOES NOT INCLUDE THE BETA SOBOLEVS)
+        beta_sobolev : Pandas DataFrame
+            Beta Sobolev factors.
+        Returns
+        -------
+        numpy.array (number of levels, number of levels)
+            Matrix with excitation-deexcitation rates(should be added to NLTE rate matrix for excitation treatment).
+            NOTE: only works with ONE ion number treated in NLTE excitation AT ONCE.
+        """
+        number_of_shells = beta_sobolev.shape[1]
```
[the remark this answers is not in the document]
Comment (sonachitchyan) [reply tardis/plasma/properties/nlte_rate_equation_solver.py:750]: I can leave it for now, and when we have the full implementation, see if switching it slows down the code

[the remark this answers is not in the document]
Comment (sonachitchyan) [reply tardis/plasma/properties/nlte_rate_equation_solver.py:753]: make an issue

**Review by Rodot- — APPROVED** [review]

**Review by chvogl — COMMENTED** [review]
Looks great.

File: tardis/plasma/properties/nlte_rate_equation_solver.py [hunk 24]
```diff
@@ -699,3 +712,140 @@ def prepare_solution_vector(self, number_density):
...
+        r_lu_matrix = np.zeros_like(r_ul_matrix)
+        r_lu_matrix_reshaped = r_lu_matrix.reshape(
+            (number_of_levels**2, number_of_shells)
+        )
+        r_lu_matrix_reshaped[r_lu_index] = (
+            B_lus[np.newaxis].T * j_blues_filtered
+        )
+        return (
+            lines_index,
+            number_of_levels,
```
[the remark this answers is not in the document]
Comment (sonachitchyan) [reply tardis/plasma/properties/nlte_rate_equation_solver.py:845]: yeah that was a mistake.

**Review by chvogl — APPROVED** [review]

**Review by Rodot- — APPROVED** [review]
